### ISSUE TITLE
Ensure that AppImages with spaces in the filename work

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#678]
Continuation from https://github.com/nomad-desktop/nx-software-center/issues/28

Needs to be investigated whether AppImageKit is at fault or linuxdeployqt or something else.

**Comment by TheAssassin** [comment]
Just another "missing sanitation" error in our code base...

**Comment by KD0BPV** [comment]
I think this or a similar issue is also affecting the AppRun binary. I'll list various Exec lines I tried in my .desktop file, and AppRun's error regarding it. The script file referenced is at `"//usr/bin/Katawa Shoujo.sh"`.

`Exec="Katawa Shoujo.sh"` > `Error executing '"Katawa Shoujo.sh"': No such file or directory`
`Exec='Katawa Shoujo.sh'` > `Error executing ''Katawa': No such file or directory`
`Exec=Katawa\ Shoujo.sh` > `Error executing 'Katawa\': No such file or directory`

**Comment by TheAssassin** [comment]
@KD0BPV To be fair, executable names with spaces are something totally awkward...

**Comment by KD0BPV** [comment]
No argument from me there. But then, I'm not the author of that app. lol. And since the standards involved allow it (Linux kernel, FreeDesktop .desktop files, desktop environments, etc), it needs to be handled. 

In the end, I patched the shell script to properly launch the python environment that the app needs, regardless of the script's filename, and changed the Exec line and filename accordingly. My AppImage works perfectly after that. :smile:

**Comment by TheAssassin** [comment]
Didn't try to excuse this bug at all. Just a comment about antipatterns when choosing executable filenames...

**Comment by probonopd** [comment]
AppImageKit _wants to_ support file names with spaces in them. If we currently don't, then it's a bug.

**Comment by KurtPfeifle** [comment]
@probonopd:

> "AppImageKit *wants* to support file names with spaces in them. If we currently don't, then it's a (one or more) bug."

The [AppImageSpec draft document]() explicitly *demands* tolerance for spaces in file system paths as well as in AppImage filenames, for both, type 1 and type 2 AppImages:

> **MUST** work even when stored in a filesystem path that contains blanks or when stored with a file name that contains blanks

**Comment by TheAssassin** [comment]
@KurtPfeifle jfyi, that part is apparently discussing the file path of the AppImage, not paths _within_ the AppImage.

**Comment by probonopd** [comment]
We want to support spaces in paths both outside and inside of the AppImage. For the latter, `AppRun` also needs to be coded accordingly. I _hope_ ours are - but we need to check this.

**Comment by TheAssassin** [comment]
I assume the problem is the lack of escaping of the `Exec=<filename>` filename part. I haven't got to looking into this yet.

**Comment by probonopd** [comment]
Good point. It _may_ be necessary to use `Exec=sh -c "Katawa Shoujo.sh"`.

References:
* https://askubuntu.com/questions/189822/how-to-escape-spaces-in-desktop-files-exec-line
* https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s07.html

**Comment by probonopd** [comment]
But https://github.com/nomad-desktop/nx-software-center/issues/28 seems to be about spaces in the name of the AppImage itself.

**Comment by KD0BPV** [comment]
As the original author of  nomad-desktop/nx-software-center#28, yes. That is what we discovered in that issue: An .AppImage file with spaces in the name does not get it's filename properly handled. In that case, it caused the run-time  to believe the mount failed, because it was looking for in the wrong (non-existing) mount point.

I did some testing, and I think you're right about the `sh -c "..."` point. The .desktop files don't appear to handle spaces, quotes, and escaped characters the same way that shells do, so the AppRun issue I was having may have been my own fault. I'll test to confirm, and let you know the results.

**Comment by KD0BPV** [comment]
Ok, yeah. There does *NOT* appear to be a problem with the AppRun binary after all. I removed my patch from `Katawa Shoujo.sh`, renamed it accordingly, and changed the .desktop to `Exec=sh -c "Katawa Shoujo.sh"`, and now the game launches as expected.

**Comment by KurtPfeifle** [comment]
It would be really ridiculous to demand support for paths and file names with spaces only for places outside the AppImage, but not for paths inside the squashfs file system. If the AppImageSpec leaves room for misinterpretation here, it should be fixed.

**Comment by KD0BPV** [comment]
I was wrong, there is an issue with AppRun, but not what I originally thought. Please see #690.

**Comment by TheAssassin** [comment]
@KurtPfeifle I agree. Please send a PR.

**Comment by TheAssassin** [comment]
Related issue: #706.

**Comment by rdiez** [comment]
I am having trouble starting the following KiCad AppImage:  https://downloads.kicad.org/kicad/linux/explore/stable/download/kicad-10.0.3-x86_64.AppImage.tar

So I unpacked it with `--appimage-extract`. The path to the unpacked directory had spaces in the name, and I got the following errors:

```
ERROR: ld.so: object '/home/rdiez/rdiez/LocalSoftware/KiCad/unpacked' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object 'AppImage/AppDir/shared/lib/jsc-stack-fix.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
```
There is a space after "unpacked" in the directory name, and that is where the path got cut off. As soon as I removed the spaces in the path, the error messages went away.

I wonder where the problem is. Is that an issue in the way KiCad has created the AppImage? Or does AppRun have problems when the path has spaces?

I found some oldish GitHub issues about spaces in the paths, but I couldn't find a clear match, or an indication whether this problem existed and has been fixed. This is the closest bug report I found.

**Comment by TheAssassin** [comment]
That AppImage is using a third-party setup. Please report your issue there. It's not an issue of AppImage's tooling.

**Comment by rdiez** [comment]
Thanks for looking. I have reported it to the KiCad project here: https://gitlab.com/kicad/code/kicad/-/work_items/24775
